replacy/__init__.py

```diff
@@ -149,64 +154,156 @@ def get_predicates(self, match_hooks) -> List[Callable]:
-                    # so, per python conventions, PATTERN_REF = -1 would mean the last matched token
-                    # so we can just add ref and end if ref is negative
-                    refd_token = doc[end + ref]
-            changed_text = None
-
-            # check if inflect
-            if "FROM_TEMPLATE_ID" in item:
-                template_id = item["FROM_TEMPLATE_ID"]
-                index = None
-                for i, token in enumerate(self.match_dict[match_name]["patterns"]):
-                    if "TEMPLATE_ID" in token and token["TEMPLATE_ID"] == template_id:
-                        index = i
-                        break
-                if index is not None:
-                    changed_text = self.inflector.auto_inflect(doc, text, start + index)
-            elif "PATTERN_REF" in item:
-                if "INFLECTION" in item:
-                    form = item["INFLECTION"]
-                    changed_text = self.inflector.inflect_or_lookup(refd_token, form)
-                else:
-                    changed_text = refd_token.text
-
-                # This should probably be a list of ops
-                # and we should have a parser class
-                if "REPLACY_OP" in item:
-                    op = item["REPLACY_OP"]
-                    if op == "LOWER":
-                        changed_text = changed_text.lower()
-                    if op == "TITLE":
-                        changed_text = changed_text.title()
-                    if op == "UPPER":
-                        changed_text = changed_text.upper()
-            if changed_text:
-                text_list.append(changed_text)
-            elif len(text):
-                text_list.append(text)
-        return " ".join(text_list)
+
+            item_options = self.get_item(item, doc, start, end)
+            inflected_options = self.inflect_item(
+                item_options, item, doc, start, end, match_name
+            )
+            cased_options = self.case_item(inflected_options, item)
+            options.append(cased_options)
+
+        # assert there aren't more than max_suggestions_count
+        # otherwise raise warning and return []
+        suggestions_count = (
+            seq(options).map(lambda x: len(x)).reduce(lambda x, y: x * y)
+        )
+
+        if suggestions_count > self.max_suggestions_count:
+            warnings.warn(
+                f"Got {suggestions_count} suggestions, max is {self.max_suggestions_count}. \
+                Will fallback to empty suggestions."
+            )
+            opt_text = []
+        else:
+            opt_combinations = list(itertools.product(*options))
+            opt_text = [" ".join(list(o)) for o in opt_combinations]
+
+        return opt_text
+
+    def score_suggestion(self, doc, span, suggestion):
+        text = " ".join([doc[: span.start].text, suggestion, doc[span.end :].text])
+        return self.scorer(text)
```

Comment:
Sorry, I don't really know if my code samples are right, since I don't know the type signatures

`self.scorer` is `None` if there is no `KenLM`... Should we have the line that sets it be :

```python
self.scorer = (
            KenLMScorer(nlp=self.nlp, model=load_lm(lm_path)) if lm_path else lambda x: 1
        )
```

e.g. set it to the constant function?

Or, if there is more than 1 suggestion, maybe `sort_suggestions` should have behavior that depends on `self.scorer`... meaning "if it's None, either don't change the order, or randomize it"

Reply:
Can you reproduce the bug?

Sorting happens only if there is the scorer, otherwise this step is completely skipped:
https://github.com/Qordobacode/replaCy/blob/c6eb41fba0b430c95b1feee5261807df716b3151/replacy/__init__.py#L392

I think I disagree with randomizing suggestions since this doesn't look like a desired behavior before adding `kenlm` - what if users want to manually set the order of suggestions (because they feel some options will be better)?

Reply:
Take a look at `tests/test_suggestions.py`. This is the case when there is no `lm_path`.